### PR TITLE
Fix the race condition caused by `with_new_connections_blocked`

### PR DESCRIPTION
**Pull request by kamipo**
`with_new_connections_blocked` was introduced at #14938.
But the method sometimes causes `@new_cons_enabled = false` then never
toggled to true.

Related https://github.com/sonots/activerecord-refresh_connection/issues/5#issuecomment-231571908.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
r? @matthewd 

cc @thedarkone

**Comment by thedarkone** [comment]
@kamipo is right - indeed, there is a bug.

Re the code, I think we should get by on a single i-var `@new_conn_blocks` (and drop `@new_cons_enabled` altogether).

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -700,13 +699,10 @@ def checkout_for_exclusive_access(checkout_timeout)
...
         def with_new_connections_blocked
-          previous_value = nil
-          synchronize do
-            previous_value, @new_cons_enabled = @new_cons_enabled, false
-          end
+          synchronize { @new_cons_blocks += 1 }
           yield
         ensure
-          synchronize { @new_cons_enabled = previous_value }
+          synchronize { @new_cons_blocks -= 1 }
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: @thedarkone should we be servicing `@available` here, too? (if we just got to zero)

AFAICS, if a thread was blocked, it's now [waiting](https://github.com/kamipo/rails/blob/700639e4f6d4d9e9be50fe282756045f9189dd23/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L730), but no-one is arranging to make the connection for it -- it's just left to hope some other already-established connection will be returned in time.

**Comment by thedarkone** [comment]
@matthewd yes, you are right. The worst example would be an empty pool, (1) thread A temporarily blocks new conns and in the meantime (2) thread B tries to acquire a conn, is disallowed to make new ones, (3) thread B goes into queue and will get stuck there until some other thread attempts to help it, (4) thread A exits `with_new_connections_blocked` and leaves thread B stucked...

However, `with_new_connections_blocked` is an internal private method that is used in 2 places:     
 1. `disconnect` where after a forceful `conn.steal!; conn.disconnect` the pool does nothing for the threads waiting in a queue (no wake up call, no nothing), so we're consistently bad at treating those waiting threads
 2. `clear_reloadable_connections` takes care of conns waiting in a queue, so there shouldn't be a problem, right?

**Comment by matthewd** [comment]
Ah, indeed. Maybe we just need to move that "hand out connections again" code from `clear_reloadable_connections` to `with_new_connections_blocked`?

I guess one could argue that a `disconnect` method that goes out of its way to establish new connections before it returns is a bit questionable... but the alternative doesn't seem desirable, and once disconnect leaves its synchronize block, other threads would be within their rights to connect anyway -- we're really just helping them along a little.

In fact, if someone really wanted "disconnect and stay disconnected while I do this thing", I guess that would mean they needed to wrap the disconnect + thing with a call to `with_new_connections_blocked`. If someone has a compelling use case for that, we might want to make the latter public. We can cross that bridge when we come to it, though.

**Comment by thedarkone** [comment]
> Ah, indeed. Maybe we just need to move that "hand out connections again" code from `clear_reloadable_connections` to `with_new_connections_blocked`?

Agreed.

Re `disconnect` reasoning/behavior, also agreed. The pool's `disconnect` API (as judged by the long standing implementation) is not meant to be fatal/irrevocable.

**Comment by matthewd** [comment]
Backported as fd21a5736b857a0f38c2617bf6e83515aeee755a

**Pull request merged**